### PR TITLE
Re-export `Foldable1`/`Bifoldable` from `base`, migrate instances

### LINKED ISSUE #130: Re-export `Foldable1` and `Bifoldable1` from `base-4.18` (GHC 9.6)
`base-4.18.0.0` (bundled with GHC 9.6) implements [this CLC proposal](https://github.com/haskell/core-libraries-committee/issues/9) (see [this GHC commit](https://gitlab.haskell.org/ghc/ghc/-/commit/58530271e5e3044623b7e2343f2152ca0092b5a2)), which adds `semigroupoids`' `Foldable1` and `Bifoldable1` classes to `base`. We should re-export these classes when building with a sufficiently new version of `base`, and provide backported versions of these classes when building with older versions of `base`. Note that the API of these classes in `base` have changed somewhat from their current form in `semigroupoids`, so we will need to make some API changes on the `semigroupoids` end.

After doing this, we can close the following issues, as they are now `base` issues rather than `semigroupoids` issues:

* #49
* #53
* #55
* #77

It would also resolve some parts of #26, but there are certainly other parts of `semigroupoids` to bikeshed besides `Foldable1` and `Bifoldable1`.

### PR DESCRIPTION
**Pull request by RyanGlScott**
This patch:

* Re-exports the `Foldable1` and `Bifoldable` classes from `base` or `foldable-classes-compat`, depending on which version of `base` is being used.
* Migrates the `Foldable1` and `Bifoldable1` instances to their respective downstream packages. I have bumped the lower version bounds on `bifunctors` and `tagged` to ensure that they always provide these instances.

Fixes #130.

-----

This PR is marked as a draft pending a Hackage release of ekmett/bifunctors#113.

### CODE REVIEW

**Comment by RyanGlScott** [comment]
The GHC 9.6 build is [failing with](https://github.com/ekmett/semigroupoids/actions/runs/4109567633/jobs/7091538994#step:24:37):

```
Building library for semigroupoids-5.4..

src/Data/Semigroup/Traversable/Class.hs:57:0: error:
     error: missing binary operator before token "("
     #if defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS
     
   |
57 | #if defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS
   | ^

src/Data/Semigroup/Traversable/Class.hs:216:0: error:
     error: missing binary operator before token "("
     #if defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS
     
    |
216 | #if defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS
    | ^

src/Data/Semigroup/Traversable/Class.hs:57:0: error:
     error: missing binary operator before token "("
     #if defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS
     
   |
57 | #if defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS
   | ^

src/Data/Semigroup/Traversable/Class.hs:216:0: error:
     error: missing binary operator before token "("
     #if defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS
     
    |
216 | #if defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS
    | ^

src/Data/Semigroup/Traversable/Class.hs:1:1: error:
    `gcc' failed in phase `C pre-processor'. (Exit code: 1)
  |
1 | {-# LANGUAGE CPP, TypeOperators #-}
  | ^
Error: cabal-3.9.0.0: Failed to build semigroupoids-5.4.
```

Strangely, this doesn't happen when I build locally.

File: semigroupoids.cabal [hunk 1]
```diff
@@ -126,11 +127,15 @@ library
...
     bifunctors          >= 5.5.9   && < 6,
     template-haskell    >= 0.2.11,
     transformers        >= 0.5     && < 0.7,
     transformers-compat >= 0.6     && < 0.8
 
+  if !impl(ghc >= 9.6)
+    build-depends: foldable1-classes-compat >= 0.1 && < 0.2
+
   if flag(containers)
     build-depends: containers >= 0.5.7.1 && < 0.7
```
Comment (phadej) [inline semigroupoids.cabal:140]: Try something like

```
cabal build -w $HOME/.ghcup/bin/ghc-9.6.0.20230128 all -f-containers --allow-newer=template-haskell --allow-newer=base
```

The `-f-containers` is import bit.

But TBH, I don't know why one would disable `containers` flag ever. I'd simply remove it.

Comment (RyanGlScott) [reply semigroupoids.cabal:140]: Ah, I see. After macro expansion, `defined(MIN_VERSION_containers) && HAS_FOLDABLE1_CONTAINERS` becomes:

```c
defined(MIN_VERSION_containers) && MIN_VERSION_containers(0,6,7)
```

But if `MIN_VERSION_containers` isn't defined, then that is a malformed macro. (I had assumed that `define(MIN_VERSION_containers)` would short-circuit if it wasn't defined, but sadly, that is not the case.)

I've pushed a new version which ensures that puts the macros into a well behaved order.

> But TBH, I don't know why one would disable `containers` flag ever. I'd simply remove it.

Yes, I feel the same way. I've opened #133 to track removing this flag.

**Comment by RyanGlScott** [comment]
I am thinking that this warrants bumping the "Kmett-major" version number to `6`, as this change breaks at least one library in the Kmettiverse (`streams`):

```
[1 of 6] Compiling Data.Stream.Future ( src/Data/Stream/Future.hs, interpreted )

src/Data/Stream/Future.hs:104:10: warning: [-Wmissing-methods]
    • No explicit implementation for
        either ‘foldMap1’ or ‘foldrMap1’
    • In the instance declaration for ‘Foldable1 Future’
    |
104 | instance Foldable1 Future
    |          ^^^^^^^^^^^^^^^^
[2 of 6] Compiling Data.Stream.Future.Skew ( src/Data/Stream/Future/Skew.hs, interpreted )

src/Data/Stream/Future/Skew.hs:271:18: error:
    Ambiguous occurrence ‘last’
    It could refer to
       either ‘Data.Semigroup.Foldable.last’,
              imported from ‘Data.Semigroup.Foldable’ at src/Data/Stream/Future/Skew.hs:65:1-30
              (and originally defined in ‘foldable1-classes-compat-0.1:Data.Foldable1’)
           or ‘Data.Stream.Future.Skew.last’,
              defined at src/Data/Stream/Future/Skew.hs:271:1
    |
271 | last (_ :< as) = last as
    |                  ^^^^
```

**Comment by phadej** [comment]
> I am thinking that this warrants bumping

That's fair, the change is quite substantial. That said, we could omit exports of the `last`, `head`, `minimum` and `maximum` from `semigroupoids` modules as these clash with `Prelude` (on purpose), and kmettverse likes to import stuff from itself unqualified.

There is no cure for `foldMap1` really. There is no `Traversable1` in `base` to base default implementation on.

&dagger; Unfortunately https://github.com/ghc-proposals/ghc-proposals/issues/544 is not yet a thing, so you need an `-Werror=missing-methods` - which `haskell-ci` sets - to catch those. I'll appreciate if you mention that on the GHC proposal, so it's nudged forward.

**Comment by RyanGlScott** [comment]
Indeed, it seems likely that we'll still break some things no matter how carefully we control the exports. This makes me inclined to just make `semigroupoids` match the API that `Data.Foldable1` offers in `base`, rather than export an arbitrary subset of the API.

I'm afraid that I don't understand your point about `foldMap1`. Unless I'm missing something, both `semigroupoids`' current definition of `Foldable1` and the new version of `Foldable1` in `base` offer a `foldMap1` method, right?

**Comment by phadej** [comment]
> I'm afraid that I don't understand your point about foldMap1. Unless I'm missing something, both semigroupoids' current definition of Foldable1 and the new version of Foldable1 in base offer a foldMap1 method, right?

Yes they do. But in current semigroupoids we have

```haskell
foldMap1 f = maybe (error "foldMap1") id . getOptionCompat . foldMap (optionCompat . Just . f)
```

i.e. the default is defined in terms of `foldMap` (I remembered that we as defined in terms of `traverse1`, the `foldMap` definitions is quite bad).

And now if someone has

```haskell
instance Foldable1 MyContainer
```

relying on default definitions, *their code will still continue to compile*, but won't work, as `-Wmissing-methods` is merely a warning.

**Comment by phadej** [comment]
> Indeed, it seems likely that we'll still break some things no matter how carefully we control the exports. 

Sure. One can argue that people should (in some timeframe) migrate out of using `semigroupoids` modules. Yet `module Data.Semigroup.Foldable` exports additional stuff, so it cannot be deprecated either.

We should control the exports at least by explicitly listing the members of `Foldable1` class. Who knows if something will be added. That's why I'd rather *re-export* as little as possible.

**Comment by RyanGlScott** [comment]
The default implementations for `base`'s` `Foldable1` do appear to be a bit different, yes. (I'm not really sure what `traverse1` has to do with the defaults, however. As far as I can tell, the default implementations for `semigroupoids`' `Foldable1` don't make use of `traverse1`.)

I think you've convinced me that we should only re-export some of `Foldable1`'s class methods from `Data.Semigroup.Classes`. Would you be OK with the following export list?

```hs
module Data.Semigroup.Class where
  ( -- Class methods of Foldable1
    Foldable1(fold1, foldMap1, toNonEmpty)
    -- Data.Foldable1 re-exports
  , intercalate1
  , foldrM1
  , foldlM1
  
    -- Functions defined on top of class methods
  , intercalateMap1
  , traverse1_
  , for1_
  , sequenceA1_
  , foldMapDefault1
  , asum1
    -- Generic defaults
  , gfold1
  , gfoldMap1
  , gtoNonEmpty
  ) where
```

**Comment by RyanGlScott** [comment]
I've went ahead and pushed the changes in https://github.com/ekmett/semigroupoids/pull/132#issuecomment-1465254712. I've also added a section to the `CHANGELOG` that discusses how one might migrate existing `Foldable1` instances that omit implementations of `foldMap1`.

@phadej, if everything looks good on your end, I'll cut new `bifunctors` and `semigroupoids` releases.

**Review by phadej — APPROVED** [review]
> if everything looks good on your end, I'll cut new bifunctors and semigroupoids releases.

:+1:

**Pull request merged**